### PR TITLE
[MLIR][LLVM] Add vararg support in LLVM::CallOp

### PR DESCRIPTION
**Pull request by ivanradanov**
Adds an attribute that specifies the callee type of a LLVMCallOp. This is needed to distinguish between vararg and non-vararg calls as currently the LLVM Dialect always translates CallOp's to non-vararg calls.

The new syntax I opted for is this:

```
%2 = llvm.call @printf vararg !llvm.func<i32 (ptr<4>, ...)> (%1, %arg0) : (!llvm.ptr<4>, i32) -> i32
```

for vararg calls, and for non-vararg calls, the function type is omitted and the type will be inferred from the operation operand and result types, for example:

```
%2 = llvm.call @foo(%1, %arg0) : (!llvm.ptr<4>, i32) -> i32
```

I think it makes sense to keep the type at the end after the `:` which describes the type as a MLIR operation and the function type separate.

If this syntax looks good I will update the tests to reflect this change.

### CODE REVIEW

**Comment by ivanradanov** [comment]
@ftynse @wsmoses

**Review by gysit — COMMENTED** [review]
Thanks for working on this.

I think personally I would put the vararg type after the arguments:
```
%2 = llvm.call @printf(%1, %arg0) vararg(!llvm.func<i32 (ptr<4>, ...)>) : (!llvm.ptr<4>, i32) -> i32
```
and wrap it into parentheses to emphasis stuff that is connected. The main advantage of this syntax is that the callee symbol and its arguments remain closely connected. What do you think? 

Overall this is mostly a matter of taste though.

**Review by ftynse — CHANGES_REQUESTED** [review]
Repeating some types feels a bit tautological, but I don't have a better suggestion for the syntax at this point. Requesting changes due to a memory problem in the code.

**Comment by ivanradanov** [comment]
Sorry I think I needed to rebase and force push because the code_formatter runner was getting confused?

Anyways, @gysit I liked your suggestion for the syntax so I have implemented that - also I did the same changes to InvokeOp and fixed the llvm<->mlir module translation and import and added a test.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1552]: NIT: As above. Also here, we should consider reusing the logic eventually (doesn't need to be this revision).

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1564]:
```suggestion
                                      /*isVarArg=*/false))
```

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp
Comment (Dinistro) [inline mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:215]: NIT: `has_value()` is not necessary
```suggestion
      if (callOp.getCalleeType())
```

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (Dinistro) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1438]:
```suggestion
          loc, funcTy, /*callee=*/nullptr, operands, directNormalDest,
```

**Review by Dinistro — COMMENTED** [review]
Thanks for pushing on this. I dropped a bunch of nit comments and questions.
I think the design is reasonable, but I'm not yet sure if your code is fully correct.

**Review by gysit — COMMENTED** [review]
I like the changes you did to the printer and parser :).

Can you add some more tests? I think it would make sense to have a roundtrip test to check the printing and parsing in mlir/test/Dialect/LLVMIR/func.mlir. Additionally, there should be a test that the translation to LLVM IR works as well.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (ftynse) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1205]: Also nit: don't start error messages with a capital.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (ftynse) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1383]: IMO, explicit `.failed()` makes this more readable than the "true-means-error" convention that was originally borrowed from LLVM and is only used in MLIR parsers.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (ivanradanov) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1325]: It is required if the callee is vararg, however, it can still be present for a non-vararg call, and we do not need (or want) to print it if the callee is non-vararg as the type is the same as the mlir op type printed after the `:`.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (ivanradanov) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1038]:
```
tools/mlir/include/mlir/Dialect/LLVMIR/LLVMTypes.h.inc:108:27: note: candidate function not viable: no known conversion from 'mlir::ValueRange::type_range' (aka 'ValueTypeRange<mlir::ValueRange>') to 'ArrayRef<mlir::Type>' for 2nd argument
  static LLVMFunctionType get(Type result, ArrayRef<Type> arguments, bool isVarArg = false);
```
I am getting this hence I added the `to_vector`

Comment (ftynse) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1038]: Yeah, it can be converted to `TypeRange` but not to `ArrayRef<Type>`. The caller needs to be updated to take `TypeRange` instead, but it's out of scope for this PR.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (ivanradanov) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1022]: I agree but I opted not to change the existing builders so as not to break code using them and adding even more builders started looking like way too many.

**Comment by ivanradanov** [comment]
Addressed most concerns and added mlir->llvm test.

I still need to add a roundtrip test

**Comment by ivanradanov** [comment]
Done with suggested changes and edited the op description as well - please let me know if there are any other concerns

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:615]:
```suggestion
    // Direct variadic call.
```

**Review by gysit — COMMENTED** [review]
Thanks for the updates.

Can you test the invoke as well? I know it is almost the same logic but it may still be nice to have test coverage.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (ivanradanov) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1325]: Oh you are indeed correct, we do not need to set that.

**Comment by ivanradanov** [comment]
Thank you for the comments, I have added invoke tests as well.

File: mlir/test/Dialect/LLVMIR/roundtrip.mlir
Comment (gysit) [inline mlir/test/Dialect/LLVMIR/roundtrip.mlir:446]: nit: FUNCV -> FUNC?

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir.mlir:1583]:
```suggestion
```
nit: i would probably drop that since it seems not relevant for this test?

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir.mlir:1577]:
```suggestion
```
nit: I would drop that since it seems not relevant for this test?

**Review by gysit — APPROVED** [review]
LGTM modulo nit comments!

Please make sure the CI is green before landing. There seems to be some formatting issue at the moment.

**Review by Dinistro — APPROVED** [review]
The additional member function seems helpful. For consistency, I would also add it to invoke.

**Comment by Dinistro** [comment]
Thank you very much for addressing all our comments and for working on extending the LLVM dialect 😄

**Pull request merged**
